src/lexer.coffee

```diff
@@ -762,11 +763,50 @@ exports.Lexer = class Lexer
                '**', 'SHIFT', 'RELATION', 'COMPARE', '&', '^', '|', '&&', '||',
                'BIN?', 'THROW', 'EXTENDS']
 
-  formatString: (str) ->
-    str.replace STRING_OMIT, '$1'
+  formatString: (str, options) ->
+    @replaceUnicodeCodePointEscapes str.replace(STRING_OMIT, '$1'), options
 
   formatHeregex: (str) ->
-    str.replace HEREGEX_OMIT, '$1$2'
+    @formatRegex str.replace(HEREGEX_OMIT, '$1$2'), delimiter: '///'
+
+  formatRegex: (str, options) ->
+    @replaceUnicodeCodePointEscapes str, options
+
+  unicodeCodePointToUnicodeEscapes: (codePoint) ->
+    toUnicodeEscape = (val) ->
+      str = val.toString 16
+      "\\u#{repeat '0', 4 - str.length}#{str}"
+    return toUnicodeEscape(codePoint) if codePoint < 0x10000
+    # surrogate pair
+    high = Math.floor((codePoint - 0x10000) / 0x400) + 0xD800
+    low = (codePoint - 0x10000) % 0x400 + 0xDC00
+    "#{toUnicodeEscape(high)}#{toUnicodeEscape(low)}"
+
+  # Replace \u{...} with \uxxxx[\uxxxx] in strings and regexes
+  replaceUnicodeCodePointEscapes: (str, options) ->
```

Comment:
Found a bug that I hadn't tested for: adjacent code point escapes eg `"\u{a}\u{b}"` were not both being processed because of the "before" section of `UNICODE_CODE_POINT_ESCAPE` (`( (?:^|[^\\]) (?:\\\\)* )`, which I copied from `STRING_INVALID_ESCAPE`). Basically, since it looks for a non-zero-length (unless we're at the beginning of the string) verification that we're not inside an "escaped escape", the desired matches eg `\u{a}` and `\u{b}` become overlapping (because the second match would be `}\u{b}`). I couldn't figure out how to make the regex not have to overlap so that I could still use `str.replace()` so I switched to an (uglier) implementation using `UNICODE_CODE_POINT_ESCAPE.exec()` and some `lastIndex` tweaking

In `test/regexps.coffee` and `test/strings.coffee`, I copied the `toJS()` helper from `test/modules.coffee` so that I could add some rewriting tests on the generated JS (since code-point escapes and regular Unicode escapes seem indistinguishable to eg `eq()`)